reco_utils/evaluation/python_evaluation.py

```diff
@@ -325,63 +352,48 @@ def merge_ranking_true_pred(
+        relevancy_method (str): method for determining relevancy ['top_k', 'by_threshold']
+        k (int): number of top k items per user (optional)
+        threshold (float): threshold of top items per user (optional)
     Returns:
-        pd.DataFrame: new data frame of true data DataFrame of recommendation hits
-            number of common users
+        pd.DataFrame, pd.DataFrame: DataFrame of recommendation hits, and hit counts vs actual relevant items per user
     """
-    relevant_func = {"top_k": get_top_k_items}
 
-    rating_pred_new = (
-        relevant_func[relevancy_method](
-            dataframe=rating_pred,
-            col_user=col_user,
-            col_rating=col_prediction,
-            threshold=threshold,
-        )
-        if relevancy_method == "by_threshold"
-        else relevant_func[relevancy_method](
-            dataframe=rating_pred, col_user=col_user, col_rating=col_prediction, k=k
-        )
-    )
+    # Make sure the prediction and true data frames have the same set of users
+    common_users = list(set(rating_true[col_user]).intersection(set(rating_pred[col_user])))
+    rating_true_common = rating_true[rating_true[col_user].isin(common_users)]
+    rating_pred_common = rating_pred[rating_pred[col_user].isin(common_users)]
 
-    common_user_list = list(
-        set(rating_true[col_user]).intersection(set(rating_pred_new[col_user]))
-    )
-
-    # To make sure the prediction and true data frames have the same set of users.
-    rating_pred_new = rating_pred_new[rating_pred_new[col_user].isin(common_user_list)]
-    rating_true_new = rating_true[rating_true[col_user].isin(common_user_list)]
-
-    n_users = len(common_user_list)
-
-    # Return hit items in prediction data frame with ranking information. This
-    # is used for calculating
-    # NDCG and MAP.
-    # Use first to generate unique ranking values for each item. This is to align with the
-    # implementation in
-    # Spark evaluation metrics, where index of each recommended items (the indices are unique
-    #  to items) is used
+    # Return hit items in prediction data frame with ranking information. This is used for calculating NDCG and MAP.
+    # Use first to generate unique ranking values for each item. This is to align with the implementation in
+    # Spark evaluation metrics, where index of each recommended items (the indices are unique to items) is used
     # to calculate penalized precision of the ordered items.
-    df_rating_pred = rating_pred_new.copy()
-    df_rating_pred["ranking"] = rating_pred_new.groupby(col_user)[col_prediction].rank(
-        method="first", ascending=False
+    if relevancy_method == 'top_k':
+        top_k = k
+    elif relevancy_method == 'by_threshold':
+        top_k = threshold
+    else:
+        raise NotImplementedError('Invalid relevancy_method')
+    df_hit = get_top_k_items(dataframe=rating_pred_common, col_rating=col_prediction, k=top_k)
```

Comment:
have you tested if this refactor is faster or profiled the functions?

Reply:
i didn't notice significant changes in speed when running the notebook unit tests, i think the only way to get there is to cache the merging functions, or to change how they are being used so that the merging only happens once. I think going for the former is the least disruptive change, although it's a bit more challenging technically

Reply:
mainly this is to improve readability / cleanliness

Reply:
awesome